universal-application-tool-0.0.1/app/services/applicant/question/DateQuestion.java

```diff
@@ -0,0 +1,122 @@
+
+  private final ApplicantQuestion applicantQuestion;
+  private Optional<Long> dayValue;
+  private Optional<Long> monthValue;
+  private Optional<Long> yearValue;
+
+  public DateQuestion(ApplicantQuestion applicantQuestion) {
+    this.applicantQuestion = applicantQuestion;
+    assertQuestionType();
+  }
+
+  @Override
+  public boolean hasQuestionErrors() {
+    return !getQuestionErrors().isEmpty();
+  }
+
+  @Override
+  public ImmutableSet<ValidationErrorMessage> getQuestionErrors() {
+    if (!isAnswered()) {
+      return ImmutableSet.of();
+    }
+    return ImmutableSet.of();
+  }
+
+  @Override
+  public boolean hasTypeSpecificErrors() {
+    return !getAllTypeSpecificErrors().isEmpty();
+  }
+
+  @Override
+  public ImmutableSet<ValidationErrorMessage> getAllTypeSpecificErrors() {
+    // There are no inherent requirements in a date question.
+    return ImmutableSet.of();
+  }
+
+  @Override
+  public boolean isAnswered() {
+    return applicantQuestion.getApplicantData().hasPath(getDayPath())
+        && applicantQuestion.getApplicantData().hasPath(getMonthPath())
+        && applicantQuestion.getApplicantData().hasPath(getYearPath());
+  }
+
+  public Optional<Long> getDayValue() {
+    if (dayValue != null) {
+      return dayValue;
+    }
+
+    dayValue = applicantQuestion.getApplicantData().readLong(getDayPath());
+
+    return dayValue;
+  }
+
+  public Optional<Long> getMonthValue() {
+    if (monthValue != null) {
+      return monthValue;
+    }
+
+    monthValue = applicantQuestion.getApplicantData().readLong(getMonthPath());
+
+    return monthValue;
+  }
+
+  public Optional<Long> getYearValue() {
+    if (yearValue != null) {
+      return yearValue;
+    }
+
+    yearValue = applicantQuestion.getApplicantData().readLong(getYearPath());
+
+    return yearValue;
+  }
+
+  public void assertQuestionType() {
+    if (!applicantQuestion.getType().equals(QuestionType.DATE)) {
+      throw new RuntimeException(
+          String.format(
+              "Question is not a DATE question: %s (type: %s)",
+              applicantQuestion.getQuestionDefinition().getQuestionPathSegment(),
+              applicantQuestion.getQuestionDefinition().getQuestionType()));
+    }
+  }
+
+  public DateQuestionDefinition getQuestionDefinition() {
+    assertQuestionType();
+    return (DateQuestionDefinition) applicantQuestion.getQuestionDefinition();
+  }
+
+  public Path getDayPath() {
+    return applicantQuestion.getContextualizedPath().join(Scalar.DAY);
+  }
+
+  public Path getMonthPath() {
+    return applicantQuestion.getContextualizedPath().join(Scalar.MONTH);
+  }
+
+  public Path getYearPath() {
+    return applicantQuestion.getContextualizedPath().join(Scalar.YEAR);
+  }
+
+  @Override
+  public String getAnswerString() {
+    String[] parts = {
+      getDayValue().map(Object::toString).orElse("-"),
+      getMonthValue().map(Object::toString).orElse("-"),
+      getYearValue().map(Object::toString).orElse("-")
```

Comment:
suggestion (blocking): could we document somewhere that we use day-month-year format?